source/tutorials/nixos-configuration-testing.md

```diff
@@ -0,0 +1,485 @@
+(integration-testing-vms)=
+
+# NixOS Configuration testing
+
+Draft 2022.08.11
+
+## What will you learn?
+
+This guide introduces the functionality of Nix Package Manager to write automated tests to debug NixOS configurations independent of a working NixOS installation.
+
+## What do you need?
+<!-- todo links -->
+
+- Basic knowledge of the Nix language.
+- A working installation of Nix Package Manager or NixOS.
+- nixpkgs path set in `$NIX_PATH`
+- NixOS configuration in vm tutorial
+
+## Automated testing of a NixOS configuration
+
+Nixpkgs provide a [test environment](https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests) to automate testing of NixOS.
+The function `nixosTest` can be used to build and run one or more virtual machines and automatically starts a script written in Python to test the NixOS configuration.
+It takes one attribute set as its argument.
+The attribute set has two attributes `nodes` and `testScript`.
+The attribute nodes contains a set of named configurations.
+This is the case because test scripts can involve more than one virtual machine.
+
+### minimal example
+
+A minimal test that can be run on the default configuration is to check if the user `root` and `alice` have access to firefox.
+
+Because in our example we only use one virtual machine we simply call it `machine` and place it inside the nodes attribute set.
+
+```nix
+let
+  nixpkgs = <nixpkgs>;
+  pkgs = import nixpkgs {};
+in
+  pkgs.nixosTest {
+    nodes.machine = { config, pkgs, ... }: {
+      ...
+    }
+  }
+```
+<!-- does it make sense to have a un-compilable section? -->
```

Comment:
````suggestion
```nix
let
  nixpkgs = <nixpkgs>;
  pkgs = import nixpkgs {};
in
  pkgs.nixosTest {
    nodes.machine = { config, pkgs, ... }: {
      # ...
    }
  }
```
````